### PR TITLE
Manage the number of concurrent gocb calls

### PR DESCRIPTION
**Pull request by adamcfraser**
Updates gocb bucket to put a limit the number of concurrent gocb calls, to avoid queue overflow errors.  This includes batching of large bulk operations.

Also fixes CPU issue related to array allocation in sharded clocks - allocation of these arrays was showing up as a significant source of CPU usage on latest perf runs.  Existing value will be much too large for everything other than the \* channel.

Includes gocb uptake of bulk operations fix.

### CODE REVIEW

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 1]
```diff
@@ -20,14 +20,22 @@ import (
 
 var gocbExpvars *expvar.Map
 
+const (
+	MaxConcurrentSingleOps = 1000 // Max 1000 concurrent single bucket ops
+	MaxConcurrentBulkOps   = 35   // Max 35 concurrent bulk ops
+	MaxBulkSetSize         = 100  // Maximum number of ops per bulk call
```
Comment (tleyden) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: The variable name doesn't match the comment -- is this specific to BulkSet or for any bulk op?

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 2]
```diff
@@ -82,12 +92,26 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+		gocbExpvars.Add("SingleOps", -1)
+	}()
 	gocbExpvars.Add("Get", 1)
 	casGoCB, err := bucket.Bucket.Get(k, rv)
 	return uint64(casGoCB), err
 }
 
 func (bucket CouchbaseBucketGoCB) GetRaw(k string) (rv []byte, cas uint64, err error) {
+
+	bucket.singleOps <- struct{}{}
```
Comment (tleyden) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go:110]: Can you add a comment above this block that says something like "Limit number of concurrent GoCB calls in order to avoid queue overflow issues"

File: src/github.com/couchbase/sync_gateway/base/sharded_sequence_clock.go [hunk 3]
```diff
@@ -371,13 +371,7 @@ func (p *ShardedClockPartition) getSequenceForOffset(offset uint16, size uint8)
...
 
 	return uint64(p.value[offset+1]) |
-		uint64(p.value[offset])<<8 |
-		uint64(0)<<16 |
-		uint64(0)<<24 |
-		uint64(0)<<32 |
-		uint64(0)<<40 |
-		uint64(0)<<48 |
-		uint64(0)<<56
+		uint64(p.value[offset])<<8
```
Comment (tleyden) [inline src/github.com/couchbase/sync_gateway/base/sharded_sequence_clock.go:374]: I think the previous version with the 0 placeholders was simpler to understand:

```
uint64(0)<<16 |     
        uint64(0)<<24 |     
        uint64(0)<<32 |     
        uint64(0)<<40 |     
        uint64(0)<<48 |     
        uint64(0)<<56
```

Also this change feels a bit unrelated to the rest of the changes in this PR/commit.

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go (hunk 1 above)
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: Had originally intended 'set' to mean 'collection' here, but I see that's confusing.  Have renamed to MaxBulkBatchSize.

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go (hunk 2 above)
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go:110]: Added the comment a little bit further up (in GetCouchbaseBucketGoCB).

File: src/github.com/couchbase/sync_gateway/base/sharded_sequence_clock.go (hunk 3 above)
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/sharded_sequence_clock.go:374]: Given that get/set sequence are very high frequency operation, taking the placeholders out was a performance optimization.

There's another critical performance fix in sharded_sequence_clock in this PR (the map allocation).  I agree it would be ideal as a separate PR, but left it in this one to save time.

**Comment by adamcfraser** [comment]
Jenkins dev build is clean:
http://mobile.jenkins.couchbase.com/view/Sync_Gateway_Dev/job/sync-gateway-dev-build/377/

**Comment by tleyden** [comment]
Looks good, can you squash into one commit?

**Comment by tleyden** [comment]
This PR was replaced by https://github.com/couchbase/sync_gateway/pull/1315
